src/interrupt.rs

```diff
@@ -1,16 +1,29 @@
 //! Interrupts
 
+// use core::sync::atomic::{self, Ordering};
+
 pub use bare_metal::{CriticalSection, Mutex, Nr};
 
 /// Disables all interrupts
 #[inline]
 pub fn disable() {
     match () {
-        #[cfg(target_arch = "arm")]
+        #[cfg(all(cortex_m, feature = "inline-asm"))]
         () => unsafe {
             asm!("cpsid i" ::: "memory" : "volatile");
         },
-        #[cfg(not(target_arch = "arm"))]
+
+        #[cfg(all(cortex_m, not(feature = "inline-asm")))]
+        () => unsafe {
+            extern "C" {
+                fn __cpsid();
+            }
+
+            // XXX do we need a explicit compiler barrier here?
```

Comment:
My experiments indicate that FFI calls behave the same as `asm!("" ::: "memory" : "volatile")` (I suppose because the code that's being called could be doing an assembly call with those clobbers (?)) so I think `compiler_barrier` is not necessary here.